// @grammyjs/storage-redis/dist/cjs/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisAdapter = void 0;
class RedisAdapter {
    /**
     * @constructor
     * @param {opts} Constructor options
     * @param {opts.ttl} ttl - Session time to life in SECONDS.
     * @param {opts.instance} instance - Instance of redis.
     * @param {opts.autoParseDates} autoParseDates - set to true to convert string in the json date format to date object
     */
    constructor({ instance, ttl, autoParseDates }) {
        if (instance) {
            this.redis = instance;
        }
        else {
            throw new Error('You should pass redis instance to constructor.');
        }
        this.ttl = ttl;
        this.autoParseDates = autoParseDates || false;
    }
    async read(key) {
        const session = await this.redis.get(key);
        if (session === null || session === undefined) {
            return undefined;
        }
        if (this.autoParseDates) {
            return JSON.parse(session, dateParser);
        }
        return JSON.parse(session);
    }
    async write(key, value) {
        await this.redis.set(key, JSON.stringify(value));
        if (this.ttl) {
            this.redis.expire(key, this.ttl);
        }
    }
    async delete(key) {
        await this.redis.del(key);
    }
}
exports.RedisAdapter = RedisAdapter;
const ISO_8601 = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z?$/;
const dateParser = (key, value) => {
    if (typeof value === 'string' && ISO_8601.test(value)) {
        var newValue;
        if (!value.endsWith("Z")) {
            newValue = `${value}Z`;
        }
        else
            newValue = value;
        return new Date(newValue);
    }
    return value;
};
